pkg/metrics/server.go

```diff
@@ -0,0 +1,99 @@
+package metrics
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/spf13/cobra"
+	"k8s.io/klog"
+)
+
+type Server struct {
+	http             http.Server
+	shutdownFinished chan struct{}
+}
+
+func (server *Server) ListenAndServe() {
+	err := server.http.ListenAndServe()
+
+	if errors.Is(err, http.ErrServerClosed) {
+		// Expected in case of shutdown.
+		err = nil
+	} else {
+		klog.Warning("HTTP server failed: " + err.Error())
+	}
+
+	<-server.shutdownFinished
+}
+
+func (server *Server) shutdown(waitForExit context.Context, exitTimeout time.Duration) {
+	<-waitForExit.Done()
+
+	ctx, cancel := context.WithTimeout(context.Background(), exitTimeout)
+	defer cancel()
+
+	err := server.http.Shutdown(ctx)
+	if err != nil {
+		klog.Warning("Error while shutting down HTTP server: " + err.Error())
+	} else {
+		klog.Info("HTTP server successfully shut down.")
+	}
+	close(server.shutdownFinished)
+}
+
+type Observable func()
+
+func poll(ctx context.Context, period time.Duration, observe Observable) {
+	ticker := time.NewTicker(period)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			observe()
+		}
+	}
+}
+
+type ServerConfig struct {
+	host       string
+	port       int16
+	pathPrefix string
```

Comment:
The fields on a config object should be exported.

```suggestion
	Host       string
	Port       int16
	PathPrefix string
```